### ISSUE TITLE
multi core issue 

### ISSUE THREAD

**Issue by CovertError** [moloch--/RootTheBox#296]
Hello

Today was trying to figure out why was the scoreboard taking awhile to process the data and to test it I used a big database with a lot of users while opening multiple tabs of the scoreboard summery and monitoring the CPU. What I noticed was that all the processing is happening on a single core when it comes to getting the data under the summary table. So what would happen is it would send the page without the data under the summary table and that would be done on multicore but when it comes to getting the data under the tables it would just do it on a single core. 

```
22:33:18     CPU    %usr   %nice    %sys %iowait    %irq   %soft  %steal  %guest  %gnice   %idle
22:33:19     all   10.50    0.00    0.62    0.00    0.00    0.12    0.12    0.00    0.00   88.62
22:33:19       0    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:19       1    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:19       2    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:19       3    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:19       4    4.17    0.00    1.04    0.00    0.00    0.00    0.00    0.00    0.00   94.79
22:33:19       5    0.00    0.00    0.99    0.00    0.00    0.00    0.00    0.00    0.00   99.01
22:33:19       6   77.67    0.00    2.91    0.00    0.00    0.97    0.00    0.00    0.00   18.45
22:33:19       7    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00

22:33:19     CPU    %usr   %nice    %sys %iowait    %irq   %soft  %steal  %guest  %gnice   %idle
22:33:20     all    7.38    0.00    0.62    0.00    0.00    0.00    0.00    0.00    0.00   92.00
22:33:20       0    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:20       1    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:20       2    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:20       3    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:20       4    4.04    0.00    1.01    0.00    0.00    0.00    1.01    0.00    0.00   93.94
22:33:20       5    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:20       6   53.92    0.00    2.94    0.00    0.00    0.00    0.00    0.00    0.00   43.14
22:33:20       7    0.00    0.00    0.98    0.98    0.00    0.00    0.00    0.00    0.00   98.04

22:33:20     CPU    %usr   %nice    %sys %iowait    %irq   %soft  %steal  %guest  %gnice   %idle
22:33:21     all    7.40    0.00    0.63    0.00    0.00    0.00    0.00    0.00    0.00   91.97
22:33:21       0    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:21       1    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:21       2    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:21       3    0.00    0.00    0.99    0.00    0.00    0.00    0.00    0.00    0.00   99.01
22:33:21       4    3.03    0.00    3.03    0.00    0.00    0.00    0.00    0.00    0.00   93.94
22:33:21       5    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:21       6   55.56    0.00    2.02    0.00    0.00    1.01    0.00    0.00    0.00   41.41
22:33:21       7    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00

22:33:21     CPU    %usr   %nice    %sys %iowait    %irq   %soft  %steal  %guest  %gnice   %idle
22:33:22     all   11.89    0.00    0.50    0.00    0.00    0.25    0.00    0.00    0.00   87.36
22:33:22       0    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:22       1    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:22       2    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:22       3    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:22       4    9.90    0.00    0.99    0.00    0.00    0.99    0.00    0.00    0.00   88.12
22:33:22       5    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:22       6   88.66    0.00    2.06    0.00    0.00    0.00    0.00    0.00    0.00    9.28
22:33:22       7    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00

22:33:22     CPU    %usr   %nice    %sys %iowait    %irq   %soft  %steal  %guest  %gnice   %idle
22:33:23     all    4.49    0.00    0.25    0.00    0.00    0.25    0.00    0.00    0.00   95.01
22:33:23       0    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:23       1    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:23       2    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:23       3    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:23       4    2.97    0.00    0.99    0.00    0.00    0.99    0.00    0.00    0.00   95.05
22:33:23       5    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:23       6   32.67    0.00    0.99    0.00    0.00    0.99    0.00    0.00    0.00   65.35
22:33:23       7    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00

22:33:23     CPU    %usr   %nice    %sys %iowait    %irq   %soft  %steal  %guest  %gnice   %idle
22:33:24     all    9.27    0.00    0.63    0.00    0.00    0.13    0.00    0.00    0.00   89.97
22:33:24       0    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:24       1    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:24       2    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:24       3    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:24       4    4.12    0.00    1.03    0.00    0.00    0.00    0.00    0.00    0.00   94.85
22:33:24       5    1.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00   99.00
22:33:24       6   68.00    0.00    3.00    0.00    0.00    2.00    0.00    0.00    0.00   27.00
22:33:24       7    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00

22:33:24     CPU    %usr   %nice    %sys %iowait    %irq   %soft  %steal  %guest  %gnice   %idle
22:33:25     all    5.52    0.00    0.13    0.00    0.00    0.25    0.00    0.00    0.00   94.10
22:33:25       0    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:25       1    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:25       2    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:25       3    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:25       4    3.06    0.00    0.00    0.00    0.00    1.02    0.00    0.00    0.00   95.92
22:33:25       5    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:25       6   41.18    0.00    1.96    0.00    0.00    0.98    0.00    0.00    0.00   55.88
22:33:25       7    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00

22:33:25     CPU    %usr   %nice    %sys %iowait    %irq   %soft  %steal  %guest  %gnice   %idle
22:33:26     all   10.43    0.00    0.50    0.13    0.00    0.13    0.00    0.00    0.00   88.82
22:33:26       0    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:26       1    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:26       2    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:26       3    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:26       4    7.14    0.00    1.02    0.00    0.00    0.00    0.00    0.00    0.00   91.84
22:33:26       5    0.00    0.00    0.00    1.00    0.00    0.00    0.00    0.00    0.00   99.00
22:33:26       6   76.53    0.00    3.06    0.00    0.00    1.02    0.00    0.00    0.00   19.39
22:33:26       7    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00

22:33:26     CPU    %usr   %nice    %sys %iowait    %irq   %soft  %steal  %guest  %gnice   %idle
22:33:27     all    9.36    0.00    0.37    0.00    0.00    0.25    0.00    0.00    0.00   90.01
22:33:27       0    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:27       1    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:27       2    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:27       3    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:27       4    6.12    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00   93.88
22:33:27       5    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:27       6   65.69    0.00    1.96    0.00    0.00    1.96    0.00    0.00    0.00   30.39
22:33:27       7    1.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00   99.00

22:33:27     CPU    %usr   %nice    %sys %iowait    %irq   %soft  %steal  %guest  %gnice   %idle
22:33:28     all    4.76    0.00    0.13    0.00    0.00    0.25    0.00    0.00    0.00   94.87
22:33:28       0    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:28       1    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:28       2    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:28       3    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:28       4    4.00    0.00    1.00    0.00    0.00    0.00    0.00    0.00    0.00   95.00
22:33:28       5    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:28       6   34.65    0.00    0.00    0.00    0.00    1.98    0.00    0.00    0.00   63.37
22:33:28       7    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00

22:33:28     CPU    %usr   %nice    %sys %iowait    %irq   %soft  %steal  %guest  %gnice   %idle
22:33:29     all    6.88    0.00    0.50    0.00    0.00    0.25    0.12    0.00    0.00   92.25
22:33:29       0    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:29       1    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:29       2    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:29       3    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
22:33:29       4    3.09    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00   96.91
22:33:29       5    0.99    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00   99.01
22:33:29       6   49.00    0.00    3.00    0.00    0.00    2.00    0.00    0.00    0.00   46.00
22:33:29       7    1.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00   99.00 
```

**Comment by moloch--** [comment]
This is either a limitation of your database process, or the [Python language itself](https://realpython.com/python-gil/). If you're looking for a more performant solution you can try PyPy which may help, or move the database to it's own server. Moving static assets to Nginx as a reverse proxy will also improve performance. However, largely speaking compute bound performance is not a goal of RTB.

**Comment by jeffg2k** [comment]
Yeah, I guess there are a few pieces to consider.  Is it Tornado / Python, SQL query, or client side javascript.  If it's indicative of the technologies, then to Moloch's point, not going to be able to do much about that.  But if it's inefficient coding or a poor db query or lack of index, maybe we can find a solution.  On the db side, I've heard good things about [Percona](https://www.percona.com/).  I haven't tested it with RTB, but it should be a drop in replacement for MySQL and I think it has the ability to profile.

I'll outline the process taken when loading the scoreboard.  If you can help identify the bottleneck using your load testing (maybe add some timing checks to the code), perhaps we can isolate the issue and see if anything can be done within the scope and goals of RTB.  

The initial template `scoreboard/scoreboard.html` is served up by the ScoreboardHandler class in `handler/ScoreboardHandler.py`.  The template calls scoreboard.js which makes several calls at document load that initializes a websocket for any push updates, which reruns the process.

First is `ScoreboardDataSocketHandler`, which ends up calling `libs\Scoreboard.py` now().  This could be a area for focus, in particular I see a call to `app.settings["scoreboard_history"]`.  I've run into issues in the past on scoreboard_history with performance regarding snapshots - if it is what I think, it can contain a lot of data and can quickly become huge when doing frequent snapshots with a large user base over time. Consider commenting some of this code to see what it does to load time.

After that return an AJAX call to the ScoreboardAjaxHandler class in `handler/ScoreboardHandler.py` for summary is done, which calls the template `scoreboard/summary_table.html`.   `summary_table.html` is another likely candidate for review / performance monitoring.  It has a lot of loops that query for boxes / flags / hints etc.  I also expect building this template is limited to a single processor.  Commenting code in this file might identify what may be dragging down the performance or perhaps the template could be broken into asynchronous components.

Lastly, the data is read by javascript (client side processing) in `summary.js` to update the table and provide the dynamic effects.

I think the big areas (if this is a coding bottleneck) would be `libs\Scoreboard.py` now() function, `templates/scoreboard/summary_table.html`, and `static/js/pages/scoreboard/summary.js`.

**Comment by moloch--** [comment]
Another option is to better support load balancing across multiple instances of RTB processes. However, the current WebSocket implementation would need to be re-written to support this, as I wrote the original RTB code before I appreciated how important "stateless" servers are.

**Comment by jeffg2k** [comment]
@CovertError is the site internet accessible?  Might be able to review what I can based on what is being queried / seen.

**Comment by CovertError** [comment]
Nope I'm running it locally I will share the database with you on a private repo

**Comment by CovertError** [comment]
Thanks so much for the help and thanks for the explanation and I will try to look into it more 





> Yeah, I guess there are a few pieces to consider. Is it Tornado / Python, SQL query, or client side javascript. If it's indicative of the technologies, then to Moloch's point, not going to be able to do much about that. But if it's inefficient coding or a poor db query or lack of index, maybe we can find a solution. On the db side, I've heard good things about [Percona](https://www.percona.com/). I haven't tested it with RTB, but it should be a drop in replacement for MySQL and I think it has the ability to profile.
> 
> I'll outline the process taken when loading the scoreboard. If you can help identify the bottleneck using your load testing (maybe add some timing checks to the code), perhaps we can isolate the issue and see if anything can be done within the scope and goals of RTB.
> 
> The initial template `scoreboard/scoreboard.html` is served up by the ScoreboardHandler class in `handler/ScoreboardHandler.py`. The template calls scoreboard.js which makes several calls at document load that initializes a websocket for any push updates, which reruns the process.
> 
> First is `ScoreboardDataSocketHandler`, which ends up calling `libs\Scoreboard.py` now(). This could be an area for focus, in particular I see a call to `app.settings["scoreboard_history"]`. I've run into issues in the past on scoreboard_history with performance regarding snapshots - if it is what I think, it can contain a lot of data and can quickly become huge when doing frequent snapshots with a large user base over time. Consider commenting some of this code to see what it does to load time.
> 
> After that return, an AJAX call to the ScoreboardAjaxHandler class in `handler/ScoreboardHandler.py` for summary is done, which calls the template `scoreboard/summary_table.html`. `summary_table.html` is another likely candidate for review / performance monitoring. It has a lot of loops that query for boxes / flags / hints etc. I also expect building this template is limited to a single processor. Commenting code in this file might identify what may be dragging down the performance or perhaps the template could be broken into asynchronous components.
> 
> Lastly, the data is read by javascript (client side processing) in `summary.js` to update the table and provide the dynamic effects.
> 
> I think the big areas (if this is a coding bottleneck) would be `libs\Scoreboard.py` now() function, `templates/scoreboard/summary_table.html`, and `static/js/pages/scoreboard/summary.js`.

**Comment by jeffg2k** [comment]
Ran some quick tests - it doesn't appear to be the javascript.  A refresh using 124 teams on the scoreboard took 2 seconds under single client load.

45% of the time was spent in `libs\Scoreboard.py now()`
45% of the time was spent in `templates/scoreboard/summary_table.html`
10% of the time was spent in `static/js/pages/scoreboard/summary.js` parsing

Both sections loop through the various teams, flags, hints, etc and spend about .01 sec per team to do so.  I wonder if we can use the data from Scoreboard now() to populate the summary_table so that it doesn't have to go through the process twice.  That could cut the time in half.  If we then cached it and updated in a separate thread on any event change, it could potentially reduce the other 45% by a large amount.
